### PR TITLE
DEV-2295 Pipeline job builder

### PR DESCRIPTION
**Pull request by imuchnik**
Adding support for Jenkins pipelines

### CODE REVIEW

File: example-jobs/pipeline_job.groovy [hunk 1]
```diff
@@ -0,0 +1,48 @@
+import jenkins.automation.builders.PipelineJobBuilder
+
+def script = """
```
Comment (yunghuffy) [inline example-jobs/pipeline_job.groovy]: This is cleaner when defined as stripping the top line, and not including the bottom line:

```
def script ="""\
...
}"""
```

File: example-jobs/pipeline_job.groovy [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    logRotator {
+        numToKeep(365)
+    }
+}
+
+def pipelineJobStages = new PipelineJobBuilder(
+        name: 'Pipeline builder',
+        description: 'this is a simple pipeline job',
+        stages: [[
+                         stageName: 'First stage',
```
Comment (acrewdson) [inline example-jobs/pipeline_job.groovy]: wonky indentation here

File: src/main/groovy/jenkins/automation/builders/FlowJobBuilder.groovy [hunk 3]
```diff
@@ -26,7 +26,9 @@ class FlowJobBuilder {
     List<String> emails
 
     /**
-     * @param DSL factory class, provided by Jenkins when executed from build context
+
+     * @param DSL factory class,  provided by Jenkins when executed from build context
```
Comment (acrewdson) [inline src/main/groovy/jenkins/automation/builders/FlowJobBuilder.groovy]: somehow two spaces snuck in before 'provided'

File: src/main/groovy/jenkins/automation/rest/RestApiScriptRunner.groovy [hunk 4]
```diff
@@ -3,7 +3,6 @@
 package jenkins.automation.rest
 
 import javaposse.jobdsl.dsl.DslScriptLoader
-import javaposse.jobdsl.dsl.ScriptRequest
```
Comment (acrewdson) [inline src/main/groovy/jenkins/automation/rest/RestApiScriptRunner.groovy:6]: not sure about this?

File: src/main/groovy/jenkins/automation/builders/PipelineJobBuilder.groovy [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+/**
+ * The basic PipelineJob builder
+ *
+ * @param stages list of maps describing the stages; see example linked below for the shape of the map
+ * @param name job name
+ * @param description job description
+ * @param pipelineScript optional string containing the raw pipeline script (use instead of `stages`)
+ * @param pollScmSchedule optional string in cron format to trigger builds on a scheduled interval
+ * @param emails list of email addresses to receive notifications
+ * @see <a href="https://github.com/cfpb/jenkins-automation/blob/gh-pages/docs/examples.md#TODOXXXXXXXXXXXX" target="_blank">Pipeline job builder example</a>
```
Comment (acrewdson) [inline src/main/groovy/jenkins/automation/builders/PipelineJobBuilder.groovy]: need to update the fragment identifier

File: src/main/groovy/jenkins/automation/builders/PipelineJobBuilder.groovy [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+                warningText += """
+                    ansiColor('xterm') {
+                        echo '\\033[31m! `pipelineScript` parameter has no effect because ' +
+                        '`stages` was also supplied; choose one or the other\\033[0m'
+                    }
+                """
+            }
+
+            it.description this.description
+            CommonUtils.addDefaults(delegate)
```
Comment (acrewdson) [inline src/main/groovy/jenkins/automation/builders/PipelineJobBuilder.groovy]: we have these two lines in here twice (see 34 and 35)

File: src/main/groovy/jenkins/automation/builders/PipelineJobBuilder.groovy [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+ * @param name job name
+ * @param description job description
+ * @param pipelineScript optional string containing the raw pipeline script (use instead of `stages`)
+ * @param pollScmSchedule optional string in cron format to trigger builds on a scheduled interval
+ * @param emails list of email addresses to receive notifications
+ * @see <a href="https://github.com/cfpb/jenkins-automation/blob/gh-pages/docs/examples.md#TODOXXXXXXXXXXXX" target="_blank">Pipeline job builder example</a>
+ */
+
+class PipelineJobBuilder {
+    List stages
```
Comment (acrewdson) [inline src/main/groovy/jenkins/automation/builders/PipelineJobBuilder.groovy]: I'd vote for `List<Map> stages` here

File: example-jobs/pipeline_job.groovy (hunk 1 above)
Comment (yunghuffy) [reply example-jobs/pipeline_job.groovy]: To elaborate, here's what I had in my test script

```
def script = """\
node {
  def buildSteps = [:]

  buildSteps['Build'] = {
    sh 'env | sort'
    echo 'Build step'
    sh 'uptime'
    echo '\$class'
  }

  buildSteps['Test'] = {
    sh 'echo teeeessssttttt'
    echo 'Test step'
    sh 'uptime'
  }

  parallel buildSteps
}"""
```

Which results in

image


If I have
```
def script = """
node {
  def buildSteps = [:]

  buildSteps['Build'] = {
    sh 'env | sort'
    echo 'Build step'
    sh 'uptime'
    echo '\$class'
  }

  buildSteps['Test'] = {
    sh 'echo teeeessssttttt'
    echo 'Test step'
    sh 'uptime'
  }

  parallel buildSteps
}
"""
```

Then I end up with
image

Notice the extra lines at the beginning and end.

Comment (imuchnik) [reply example-jobs/pipeline_job.groovy]: Also notice escape slash before $class.

File: src/main/groovy/jenkins/automation/rest/RestApiScriptRunner.groovy (hunk 4 above)
Comment (imuchnik) [reply src/main/groovy/jenkins/automation/rest/RestApiScriptRunner.groovy:6]: Unused import that got cleaned up by idea.

Comment (acrewdson) [reply src/main/groovy/jenkins/automation/rest/RestApiScriptRunner.groovy:6]: 👍

**Review by acrewdson — APPROVED** [review]
looks good to me 👍

File: example-jobs/pipeline_job.groovy (hunk 1 above)
Comment (imuchnik) [reply example-jobs/pipeline_job.groovy]: Yes, true, but all things considered, the functionality is the same. Slash may see a little more magical to non-groovy people, while stripIndent() is more expressive.  I am not inclined to add another commit just for that, if you don't object.

File: src/main/groovy/jenkins/automation/builders/PipelineJobBuilder.groovy [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+                    node {
+                        ${warningText}
+                        ${stagesStr}
+                    }
+                """.stripIndent()
+            }
+
+            definition {
+                cps {
+                    sandbox(false)
```
Comment (marcesher) [inline src/main/groovy/jenkins/automation/builders/PipelineJobBuilder.groovy]: Please make `sandbox` configurable

File: src/main/groovy/jenkins/automation/builders/PipelineJobBuilder.groovy [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+ * @see <a href="https://github.com/cfpb/jenkins-automation/blob/gh-pages/docs/examples.md#pipeline-builder" target="_blank">Pipeline job builder example</a>
+ */
+
+class PipelineJobBuilder {
+    List stages
+    String name
+    String description
+    String pipelineScript = ""
+    String pollScmSchedule
+    List<String> emails
```
Comment (marcesher) [inline src/main/groovy/jenkins/automation/builders/PipelineJobBuilder.groovy]: emails / extended email aren't actually an option on pipelines, so you'll need to remove this and the extended email block below

File: docs/examples.md [hunk 10]
```diff
@@ -86,6 +86,59 @@ new FlowJobBuilder(
...
+                         parameters: "[[\$class: 'StringParameterValue', name: 'foo', value: 'bar']]"
+                 ],
+                 [
+                         stageName: 'Second stage',
+                         jobName: 'Job 2',
+                 ]],
+        emails: ['jane@example.com', 'joe@example.com']
+).build(this);
+
+ppipelineJobStages.with {
```
Comment (marcesher) [inline docs/examples.md]: typo (ppipeline). Also, I think we should encourage non-named jobs and just use the `.build(this).with {...}` convention

File: docs/examples.md [hunk 11]
```diff
@@ -86,6 +86,59 @@ new FlowJobBuilder(
...
+
+).build(this);
+
+pipelineJob.with {
+    logRotator {
+        numToKeep(365)
+    }
+}
+
+def pipelineJobStages = new PipelineJobBuilder(
```
Comment (marcesher) [inline docs/examples.md]: If we're going to support this `stages` stuff then somewhere, it needs to be made clear why you'd choose this over just using pipelineScript. If I'm coming into this Builder documentation and I see two ways to do the same thing, with no explanation of the difference or why I'd choose one over the other, I'm just confused

File: src/main/groovy/jenkins/automation/builders/PipelineJobBuilder.groovy [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+ * @see <a href="https://github.com/cfpb/jenkins-automation/blob/gh-pages/docs/examples.md#pipeline-builder" target="_blank">Pipeline job builder example</a>
+ */
+
+class PipelineJobBuilder {
+    List stages
+    String name
+    String description
+    String pipelineScript = ""
+    String pollScmSchedule
+    Boolean sandboxFlag = true
```
Comment (yunghuffy) [inline src/main/groovy/jenkins/automation/builders/PipelineJobBuilder.groovy:25]: My one gripe/nitpick is that sandbox here is set to `true`. Looking through your past commits it looks like it was `false` by default. Any reason for this change? I don't think I've come across stuff we're using that expects this to be sandboxed.

Comment (acrewdson) [reply src/main/groovy/jenkins/automation/builders/PipelineJobBuilder.groovy:25]: It's a good point. I would have assumed `false` would be the default here. We don't want that?

Comment (acrewdson) [reply src/main/groovy/jenkins/automation/builders/PipelineJobBuilder.groovy:25]: I take it all back -- I think we're good here 😅  http://unethicalblogger.com/2017/08/03/donut-disable-groovy-sandbox.html

**Comment by acrewdson** [comment]
I'd vote for merging 👍  -- and squashing when merging in this case

File: src/main/groovy/jenkins/automation/builders/PipelineJobBuilder.groovy [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+                triggers {
+                    scm pollScmSchedule
+                }
+            }
+
+            if (stages) {
+                def stagesStr = stages.collect {
+                    """
+                        stage("${it.stageName}") {
+                            build job: "${it.jobName}"
```
Comment (acrewdson) [inline src/main/groovy/jenkins/automation/builders/PipelineJobBuilder.groovy:56]: I think we need a comma at the end of this line -- see https://stackoverflow.com/a/39656390

File: src/main/groovy/jenkins/automation/builders/PipelineJobBuilder.groovy (hunk 12 above)
Comment (acrewdson) [reply src/main/groovy/jenkins/automation/builders/PipelineJobBuilder.groovy:25]: just to be clear, sandboxing by default seems good to me

File: src/main/groovy/jenkins/automation/builders/PipelineJobBuilder.groovy [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+                            ${it.parameters ? 'parameters: ' + it.parameters : ''}
```
Comment (acrewdson) [inline src/main/groovy/jenkins/automation/builders/PipelineJobBuilder.groovy:57]: I think we might need to call `.inspect()` on `it.parameters` to preserve the quoting

**Review by acrewdson — DISMISSED** [review]
see comments

File: src/main/groovy/jenkins/automation/builders/PipelineJobBuilder.groovy (hunk 13 above)
Comment (yunghuffy) [reply src/main/groovy/jenkins/automation/builders/PipelineJobBuilder.groovy:56]: Confirmed that parameters require the comma

```
[BFA] Scanning build for known causes...
[BFA] No failure causes found
[BFA] Done. 0s
org.codehaus.groovy.control.MultipleCompilationErrorsException: startup failed:
WorkflowScript: 38: expecting '}', found ':' @ line 38, column 61.
   tranet-pulse-check' parameters: [boolean
                                 ^
```
from

```
    stages {
        stage('Build seed jobs'){
            steps {
                build job: 'parameterized job' parameters: [booleanParam(name: 'FOO_ME', value: true)]
            }
        }
    }
```

Comment (yunghuffy) [reply src/main/groovy/jenkins/automation/builders/PipelineJobBuilder.groovy:56]: Other thought, can we iterate over this when we actually need to use the `stages` part of this builder? I don't think we have been using this yet.

**Pull request merged**